server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java

```diff
@@ -2018,4 +2018,33 @@ sub int as round((random() + 1) * 100)
             assertThat(response.rows()[0][0]).isEqualTo(generated);
         }
     }
+
+    @Test
+    public void test_insert_on_conflict_with_non_deterministic_column() {
+        execute("""
+            CREATE TABLE tbl (
+               id int PRIMARY KEY,
+               value DOUBLE PRECISION,
+               value_text TEXT,
+               modification_date TIMESTAMP WITHOUT TIME ZONE GENERATED ALWAYS AS current_timestamp
+            )""");
+
+        ensureGreen();
+        execute("INSERT INTO tbl (id, value) " +
+            "VALUES (1, 99999) " +
+            "ON CONFLICT (id) DO UPDATE SET value = excluded.value");
+        refresh();
+
+        execute("SELECT underreplicated_shards FROM sys.health WHERE table_name = 'tbl'");
+        assertThat(response).hasRows("0"); // Used to be > 0 because of class cast exception caused by column->value mismatch in the request
+
+        // Ensure that primary and replica has the same values for generated column.
+        execute("SELECT _raw FROM tbl");
+        String raw = (String) response.rows()[0][0];
+
+        for (int i = 0; i < 10; i++) {
+            execute("SELECT _raw FROM tbl");
+            assertThat(response.rows()[0][0]).isEqualTo(raw);
+        }
```

Comment:
Selecting `_raw` because select modification_date hits some timezone problem when JDBC is used
https://github.com/crate/crate/pull/15303

Reply:
Not sure whether we need this test as I also added IndexerTest.

I fell like IndexerTest doesn't cover all interactions between Indexer/UpdateToInsert/TransportShardUpsertAction. For example, unit test relies on specific behavior of another component when creating Indexer object
https://github.com/crate/crate/pull/15309/files#diff-5679257d0b0f40d3153d406ef4477bd35e7d8ee1e5c98f0c640c95afe3c2e041R986 which I feel like not really stable.
If _raw selection is too hacky, I can use random like in test_adds_non_deterministic_sub_columns_when_root_is_not_in_targets to avoid jdbc testing issue highlighted in https://github.com/crate/crate/pull/14323